### PR TITLE
Simple Auth support

### PR DESCRIPTION
**Pull request by ktaletsk**
Adds simple credentials support as requested in #299 

**User flow**
1. User initiate git operation that may require auth (clone/push/pull)
2. Extension attempts to clone/push/pull without credentials (assuming user set up their own SSH keys) is done exactly as before
3. If git returns an error indicating insufficient credentials, extension shows a credential form for git username and password. User credentials are collected and sent to the backend.
3a. If wrong credentials are provided form is shown over and over until successful authentication / or until user clicks 'CANCEL' button on the credentials form.

If user sets their own credentials, step 3 never happen.
When cloning/pulling from public repo, step 3 never happen

**Technical details**
On the backend, API handler is overloaded to expect auth information. `Git` class methods `clone`, `push` and `pull` are updated to allow passing of username/password information to git CLI using [pexpect](https://github.com/pexpect/pexpect) library similar to how `subprocess.Popen` is used to do that without credentials.

On the frontend, couple of new forms are added. `src/components/CredentialsBox.tsx` implements the body of credentials dialog. `src/components/CommitAuthorBox.tsx` implements similar dialog, but for author information on git commit.

**Comments**
UI is still little rough around the edges, but I hope community can help me fix that.

### CODE REVIEW

**Comment by saulshanabrook** [comment]
I tried this out locally and it worked for me. I am going to give a longer review of the code as well.

Screen Shot 2019-05-07 at 4 21 01 PM

File: jupyterlab_git/git.py [hunk 1]
```diff
@@ -435,29 +482,59 @@ def checkout_all(self, top_repo_path):
...
-    def commit(self, commit_msg, top_repo_path):
+    def commit(self, commit_msg, top_repo_path, author_name=None, author_email=None):
         """
         Execute git commit <filename> command & return the result.
         """
-        my_output = subprocess.check_output(
-            ["git", "commit", "-m", commit_msg], cwd=top_repo_path
+        command = ["git", "commit", "-m", commit_msg]
+
+        if author_name != None and author_email != None:
```
Comment (saulshanabrook) [inline jupyterlab_git/git.py]: How is this related to the auth changes? Is it necessary for them?

Comment (ktaletsk) [reply jupyterlab_git/git.py]: This is related to auth, since user should be able to provide commit information without going to Terminal. This is not necessary, but without it the user flow breaks when they click commit button.

Comment (brylie) [reply jupyterlab_git/git.py]: It is important that Git commits get attributed to a specific author, particularly in multi-user repositories. How will users set the `name`  and `email` setting for this attribution?

Quick examples:
- https://help.github.com/en/articles/setting-your-username-in-git
- https://help.github.com/en/articles/setting-your-commit-email-address-in-git

**Comment by brunowego** [comment]
Hi @ktaletsk, thanks for this, I'm looking exactly for this!

File: jupyterlab_git/git.py (hunk 1 above)
Comment (ktaletsk) [reply jupyterlab_git/git.py]: @brylie In the proposed implementation each commit is attributed individually in favor of global setting:
```
git commit -c user.name <name> -c user.email <email>
```

Global option could be added as described in the links you provided. It could a check box in the interface ("Save globally"). Let me know if you think I should enable global setting or the local one is fine for now

**Comment by ktaletsk** [comment]
@saulshanabrook, is there any update on this PR? Do you have any questions regarding proposed changes? I would rebase to the latest 0.6.0, but I need to know if this PR will be considered.

File: jupyterlab_git/git.py (hunk 1 above)
Comment (saulshanabrook) [reply jupyterlab_git/git.py]: I think a global settings makes sense to have a global flag here that is seperate from this auth support. Maybe an item in the git menu that lets you edit and set your username an email in the config.

But I am still unclear, can't we include the auth support without this extract logic around the name and email? And just assume people have already set that up globally?

Comment (saulshanabrook) [reply jupyterlab_git/git.py]: I took a look at the code. I think this looks fine as is. We can always move it to setting the global later.

**Comment by ktaletsk** [comment]
I am going to resolve all the current conflicts with master shortly

**Comment by jaipreet-s** [comment]
> But I am still unclear, can't we include the auth support without this extract logic around the name and email? And just assume people have already set that up globally?

I don't think this question from Saul was addressed, and I have the same question.

**Comment by ktaletsk** [comment]
Per @saulshanabrook,

>  I think this looks fine as is. We can always move it to setting the global later.

Should I take commit author attribution out of this Auth PR and submit as a separate PR?

**Comment by saulshanabrook** [comment]
> Should I take commit author attribution out of this Auth PR and submit as a separate PR?

That would be ideal, but I understand it's more work for you. I am OK taking this as is, but @jaipreet-s would you prefer them to be split?

**Comment by ktaletsk** [comment]
@saulshanabrook @jaipreet-s I removed my implementation of the identity feature as it was added by #366. I also rebased to the latest master. Please, take a look.

File: src/widgets/gitPushPull.ts [hunk 2]
```diff
@@ -30,26 +34,30 @@ export class GitPullPushDialog extends Widget {
...
   }
 
   /**
    * Executes the relevant service API depending on the operation and handles response and errors.
    * @param currentFileBrowserPath the path to the current repo
    */
-  private executeGitApi(currentFileBrowserPath: string) {
+  private executeGitApi(
+    currentFileBrowserPath: string,
+    auth: IGitAuth = undefined
```
Comment (fcollonval) [inline src/widgets/gitPushPull.ts]:
```ts
auth?: IGitAuth
```

**Comment by ktaletsk** [comment]
Thank you @fcollonval for the review and useful suggestions. I addressed the issues in the latest commit.

File: jupyterlab_git/git.py [hunk 3]
```diff
@@ -12,6 +13,44 @@
 ALLOWED_OPTIONS = ['user.name', 'user.email']
 
 
+class git_auth_input_wrapper:
+    """
+    Helper class which is meant to replace subprocess.Popen for communicating
+    with git CLI when also sending username and password for auth
+    """
+    def __init__(self, command, cwd, env, username, password, *args, **kwargs):
+        super(git_auth_input_wrapper, self).__init__(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply jupyterlab_git/git.py]: My bad, should have noticed that

File: jupyterlab_git/git.py [hunk 4]
```diff
@@ -12,6 +13,44 @@
 ALLOWED_OPTIONS = ['user.name', 'user.email']
 
 
+class git_auth_input_wrapper:
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply jupyterlab_git/git.py]: I was not aware about that particular convention, I will make sure to use it now

File: src/components/PathHeader.tsx [hunk 5]
```diff
@@ -91,6 +93,47 @@ export class PathHeader extends React.Component<
...
+          retry ? 'Incorrect username or password.' : ''
+        ),
+        buttons: [Dialog.cancelButton(), Dialog.okButton({ label: 'OK' })]
+      });
+      retry = true;
+
+      let response = await credentialsDialog.launch();
+      credentialsDialog.dispose();
+
+      if (response.button.label === 'OK') {
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/components/PathHeader.tsx]: Agreed. I like this option better

File: src/components/CredentialsBox.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import { Widget } from '@phosphor/widgets';
+
+/**
+ * The UI for the credentials form
+ */
+export class GitCredentialsForm extends Widget {
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/components/CredentialsBox.tsx]: > Prefer setting the input nodes to local private attributes than looking for input nodes.
Yeah, that's much better.

I rewrote `GitCredentialsForm` in the similar manner to `GitAuthorForm`.

File: src/git.ts [hunk 7]
```diff
@@ -165,11 +190,19 @@ export class Git {
...
+  ): Promise<IGitPushPullResult> {
     try {
-      let response = await httpGitRequest('/git/pull', 'POST', {
+      let obj: IGitPushPull = {
         current_path: path
-      });
+      };
+      if (auth) {
+        obj.auth = auth;
+      }
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/git.ts]: Does this notation implies 
```
let obj: IGitPushPull = {
  current_path: path,
  auth=auth
};
```
?

File: jupyterlab_git/git.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fcollonval) [reply jupyterlab_git/git.py]: If you want to know more here is the official document: https://www.python.org/dev/peps/pep-0008/#id34

File: src/git.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (fcollonval) [reply src/git.ts]: This is indeed equivalent to
```ts
let obj: IGitPushPull = {
  current_path: path,
  auth: auth
}
```

**Comment by fcollonval** [comment]
@ktaletsk I forgot to point out, you need to add `pexpect` in `setup.py` https://github.com/jupyterlab/jupyterlab-git/blob/0b4339fec57e07bbb5c780f173fa09978ae68634/setup.py#L17-L20

@jaipreet-s @saulshanabrook I think this is now ready to be merged. Could you please have a look?

**Comment by jaipreet-s** [comment]
Thanks @ktaletsk for this work! I'll attempt to take a look sometime next week.

File: src/widgets/CredentialsBox.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    const text = document.createElement('span');
+    const warning = document.createElement('div');
+
+    node.className = 'jp-RedirectForm';
+    warning.className = 'jp-RedirectForm-warning';
+    text.textContent = textContent;
+    warning.textContent = warningContent;
+    this._user.placeholder = 'username';
+    this._password.placeholder = 'password';
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/widgets/CredentialsBox.tsx]: That's really good call! I am not a BitBucket or GitLab user, so I was not aware of that clone URL schema being used. Just did a quick test with the current code and private BitBucket repo and indeed BitBucket links are hanging JupyterLab and my whole browser. I will make sure to add the suggested check for the case when username is already provided.

File: src/widgets/gitClone.ts [hunk 9]
```diff
@@ -84,9 +86,41 @@ export class GitClone extends Widget {
...
-      .then(response => {
-        if (response.code !== 0) {
-          this.showErrorDialog(response.message);
+      .then(async response => {
+        let retry = false;
+        while (response.code !== 0) {
+          if (
+            response.code === 128 &&
+            (response.message.indexOf('could not read Username') >= 0 ||
+              response.message.indexOf('Invalid username or password') >= 0)
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/widgets/gitClone.ts]: That's really good call! I am not a BitBucket or GitLab user, so I was not aware of that clone URL schema being used. Just did a quick test with the current code and private BitBucket repo and indeed BitBucket links are hanging JupyterLab and my whole browser. I will make sure to add the suggested check for the case when username is already provided.

[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/widgets/gitClone.ts]: I realized my comment is related to related but different problem. If username was provided in clone URL, then we should not ask for a username at all (show different UI form) and have a different pexpect flow on the backend (only expect a password)

[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/widgets/gitClone.ts]: Good idea. I put the constant string in `src/git.ts` and then just import it in relevant places.

I decided to use `map` instead of loop to check if response message contains any of the error strings.

File: src/widgets/CredentialsBox.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * Returns the input value.
+   */
+  getValue(): IGitAuth {
+    let credentials = {
+      username: this._user.value,
+      password: this._password.value
+    };
+    return credentials;
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/widgets/CredentialsBox.tsx]: Haha, always good to remove some code

File: src/widgets/CredentialsBox.tsx [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    warningContent: string = ''
+  ) {
+    super();
+    this.node.appendChild(this.createBody(textContent, warningContent));
+  }
+
+  private createBody(textContent: string, warningContent: string): HTMLElement {
+    const node = document.createElement('div');
+    const label = document.createElement('label');
+    this._user = document.createElement('input');
```
[the remark this answers is not in the document]
Comment (ktaletsk) [reply src/widgets/CredentialsBox.tsx]: I am not sure how to make it work. 

HTML attribute `required` (https://www.w3schools.com/tags/att_input_required.asp) does not help here. It is possible to add this tag as an additional parameter, but nothing actually enforces it:
```
this._user.required = true;
this._password.required = true;
```

I also looked in other place: https://github.com/jupyterlab/jupyterlab/blob/9a69476fc76ab473e86edb156d7ff2a0bf88e6b5/packages/apputils/src/dialog.ts#L422 I think, it might be possible to use `accept` parameter to run a check on input fields to make sure they are not empty.

Any suggestions for implementation?

**Comment by ktaletsk** [comment]
@jaipreet-s @fcollonval I went through the latest review, but I couldn't find a way to make the username and password required fields. Do you have any suggestions how to implement that?

File: src/git.ts [hunk 12]
```diff
@@ -128,6 +128,31 @@ export interface IIdentity {
   email: string;
 }
 
+/**
+ * Interface for the Git Auth request.
+ */
+export interface IGitAuth {
+  username?: string;
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply src/git.ts]: Optional status removed

File: src/widgets/CredentialsBox.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (fcollonval) [reply src/widgets/CredentialsBox.tsx]: @ktalest Thanks for the research. Unfortunately as the form tag is not directly supported for `Dialog`. There are no easy way to enforce this. I'm gonna merge this and open a separate issue for that specific point.

Note: Using the `accept` method of `Dialog` could be an option. Unfortunately the dialog is customize by injecting a widget in it and not deriving from it. So it will be quite some work.

**Comment by fcollonval** [comment]
@ktaletsk Could you please rebase your branch on the master to solve the conflicts?

**Comment by ktaletsk** [comment]
@fcollonval rebased! Let's merge it

**Comment by fcollonval** [comment]
Thx a lot @ktaletsk especially for your patience to bring it through all the comments.

**Comment by ktaletsk** [comment]
Thanks @fcollonval @jaipreet-s @saulshanabrook and everyone who participated in discussions. Feels great to contribute to this extension!

**Comment by sudeshgit** [comment]
In which release this change is shipped? I can see 0.8.1 as latest (https://pypi.org/project/jupyterlab-git/)

**Comment by fcollonval** [comment]
There is no release containing this feature yet. You can watch for new
release to be notified.
